lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/hmrcKbvCri/ContractTest.java

```diff
@@ -0,0 +1,655 @@
+                .willRespondWith()
+                .status(400)
+                .toPact();
+    }
+
+    @Test
+    @PactTestFor(pactMethod = "invalidAuthCodeRequestReturns400")
+    void fetchAccessToken_whenCalledAgainstHmrcKbvCriWithInvalidAuthCode_throwsAnException(
+            MockServer mockServer) throws URISyntaxException, JOSEException {
+
+        // Arrange
+        var credentialIssuerConfig = getMockCredentialIssuerConfig(mockServer);
+
+        when(mockConfigService.getSsmParameter(ConfigurationVariable.JWT_TTL_SECONDS))
+                .thenReturn("900");
+        when(mockConfigService.getOauthCriConfig(any())).thenReturn(credentialIssuerConfig);
+        when(mockConfigService.getCriPrivateApiKey(any())).thenReturn(PRIVATE_API_KEY);
+
+        // Signature generated by jwt.io by debugging the test and getting the client assertion
+        // JWT
+        // generated by the test as mocking out the AWSKMS class inside the real signer would be
+        // painful.
+        when(mockSigner.sign(any(), any()))
+                .thenReturn(
+                        new Base64URL(
+                                "hXYrKJ_W9YItUbZxu3T63gQgScVoSMqHZ43UPfdB8im8L4d0mZPLC6BlwMJSsfjiAyU1y3c37vm-rV8kZo2uyw"));
+        when(mockSigner.supportedJWSAlgorithms()).thenReturn(Set.of(JWSAlgorithm.ES256));
+        when(mockSecureTokenHelper.generate())
+                .thenReturn("ScnF4dGXthZYXS_5k85ObEoSU04W-H3qa_p6npv2ZUY");
+
+        // We need to generate a fixed request, so we set the secure token and expiry to
+        // constant
+        // values.
+        var underTest =
+                new CriApiService(
+                        mockConfigService, mockSigner, mockSecureTokenHelper, CURRENT_TIME);
+
+        // Act
+        CriApiException exception =
+                assertThrows(
+                        CriApiException.class,
+                        () ->
+                                underTest.fetchAccessToken(
+                                        getCallbackRequest(
+                                                "dummyInvalidAuthCode", credentialIssuerConfig),
+                                        getCriOAuthSessionItem()));
+
+        // Assert
+        assertEquals("Invalid token request", exception.getErrorResponse().getMessage());
+        assertEquals(400, exception.getHttpStatusCode());
+    }
+
+    @Pact(provider = "HmrcKbvCriProvider", consumer = "IpvCoreBack")
+    public RequestResponsePact validRequestReturnsIssuedCredential(PactDslWithProvider builder)
+            throws Exception {
+        return builder.given("dummyApiKey is a valid api key")
+                .given("dummyAccessToken is a valid access token")
+                .given("test-subject is a valid subject")
+                .given("dummyHmrcKbvComponentId is a valid issuer")
+                .given("VC givenName is Mary")
+                .given("VC familyName is Watson")
+                .given("VC birthDate is 1932-02-25")
```

Comment:
Here and in the other test I don't think you've given enough information for the HMRC KBV team to generate the VC that you expect - e.g. there's no mention of the address or txn value here.

Remember that for the provider side tests these given statements are used to set up the state to generate the VC.